### PR TITLE
sg: create a test command to run e2e tests locally

### PR DESCRIPTION
**Pull request by jhchabran**
Add `sg start enterprise-e2e` to launch the local server with settings enabling to run the e2e test suite against it locally. `sg test frontend-e2e` has been patched to make it work. 

I had to swap the order in which the envs are given to `makeEnv` to allow a command set to override environment variables from the underlying run targets, which is from the user perspective the expected behavior. 

I added along the way support to fetch secrets from GCP and inject them in an env var. Those vars can still be overridden like the others, through `GH_TOKEN=foobar sg ...` 

Also added a `preamble` key to commands, so we can display some content to the user before running a command. 

image

## Test plan

Got a green e2e test run locally.

### CODE REVIEW

**Comment by jhchabran** [comment]
@valerybugakov could you check this out and give it a spin with: 

```
cd dev/sg
go run ./... start enterprise-e2e 

# in a diff terminal
cd dev/sg 
go run ./... test frontend-e2e
```

File: dev/sg/internal/run/command.go [hunk 1]
```diff
@@ -121,7 +121,7 @@ func startCmd(ctx context.Context, dir string, cmd Command, globalEnv map[string
 
 	sc.Cmd = exec.CommandContext(commandCtx, "bash", "-c", cmd.Cmd)
 	sc.Cmd.Dir = dir
-	sc.Cmd.Env = makeEnv(globalEnv, cmd.Env)
+	sc.Cmd.Env = makeEnv(cmd.Env, globalEnv)
```
Comment (mrnugget) [inline dev/sg/internal/run/command.go:124]: Hmm, what does this change? We changed the way envs work to allow people to do this:

```
SOURCEGRAPH_DOT_COM=true sg start enterprise
```

I'm fairly sure that this worked (and the docstring for `makeEnv` calls this out as a usecase). So why this change here?

File: sg.config.yaml [hunk 2]
```diff
@@ -958,7 +965,13 @@ tests:
     cmd: yarn run jest --testPathIgnorePatterns end-to-end regression integration storybook
 
   frontend-e2e:
-    cmd: yarn run test-e2e
+    cmd: |
+      echo -e "\n\n\033[0;33mWarning, sg start enterprise-e2e must be already running in a different terminal!\033[0m\n\n" && \
```
Comment (jhchabran) [inline sg.config.yaml:969]: Using `&&` enables to ctrl-C the whole thing.

File: dev/sg/internal/run/command.go (hunk 1 above)
Comment (jhchabran) [reply dev/sg/internal/run/command.go:124]: Without this, https://github.com/sourcegraph/sourcegraph/pull/34627/files#diff-bb8e3383b3d6ce6cc850d8aea2c04d5c800914ad30da787ccca0bdab8e0fc3b4R755 doesn't work because the commandset's env var gets injected before the command env. 

In concrete terms, the `EXTSVC_CONFIG` defined by `enterprise-frontend` will shadow the blank one from the commandset.

Comment (jhchabran) [reply dev/sg/internal/run/command.go:124]: If we did not patch this, teammates would have to remember to run `EXTSVC_CONFIG="" sg start` for the `sg test frontend-e2e` to work, which is not ideal.

**Review by valerybugakov — COMMENTED** [review]
A couple of notes from testing it locally:

1. `gcloud secrets versions access` fails with `ERROR: (gcloud.secrets.versions.access)`. Removing the `gcloud secrets` command from the `sg` config and providing `GH_TOKEN` token manually from the 1Password note works.
2. `BROWSER=chrome` is required for `yarn run test-e2e` and probably should be added to `frontend-e2e` `env` config.
3. The first E2E test fails because I don't have a local user with test credentials. I can create one, but maybe we can do it automatically or add a prep command that would create that user for us.

And the most important part — I was able to start E2E tests locally with a couple of commands, that's huge!

File: dev/sg/internal/run/command.go (hunk 1 above)
Comment (mrnugget) [reply dev/sg/internal/run/command.go:124]: Ah, I see! It's about the commandset & command. Got it.

**Review by mrnugget — APPROVED** [review]

File: sg.config.yaml [hunk 3]
```diff
@@ -958,7 +965,13 @@ tests:
...
+      GH_TOKEN=$(gcloud secrets versions access latest --project=sourcegraph-ci --secret="BUILDKITE_GITHUBDOTCOM_TOKEN") yarn run test-e2e
```
Comment (valerybugakov) [inline sg.config.yaml:970]: Is it possible to make this optional out of the box, so it outputs something like this?

> It seems you don't have `gcloud secrets` access. Please provide `GH_TOKEN` from the 1Password note using env variable `GH_TOKEN=123 go run ./... test frontend-e2e`

Comment (jhchabran) [reply sg.config.yaml:970]: Yeah I can workout something like this 👍

**Comment by jhchabran** [comment]
@valerybugakov ah yes, I forgot to mention the commands required to deal with that. It's a couple of another `sg` commands 😊

File: sg.config.yaml (hunk 3 above)
Comment (mrnugget) [reply sg.config.yaml:970]: Just chiming in here to say: @jhchabran don't forget that you can pull stuff into Go. You don't have to do everything inside YAML. You can add a `requiresGCloudSecret:` attribute to test cmds, for example and do the checking and command running inside Go.

Comment (jhchabran) [reply sg.config.yaml:970]: Yes, absolutely! I'll keep this as draft until it's done, at least @valerybugakov will be able to use that branch to test their stuff until then.

**Comment by jhchabran** [comment]
@valerybugakov can you try again with the last commit? If this works for you, that should be okay to unblock you for a while until I come back on this next week.

File: dev/run-frontend-e2e.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+echo -e "\n\n\033[0;33msg start enterprise-e2e must be already running in a different terminal!\033[0m" 
+echo -e "If you run into authentication issues, you can run the following commands to fix them:"
+echo -e "  sg db reset-pg && sg db add-user -name test@sourcegraph.com -password supersecurepassword\n\n"
+
+if [ -z "$GH_TOKEN" ]; then 
+  GH_TOKEN=$(gcloud secrets versions access latest --project=sourcegraph-ci --secret="BUILDKITE_GITHUBDOTCOM_TOKEN" 2>/dev/null)
+  if [ "$GH_TOKEN" == "" ]; then 
+    echo -e "GH_TOKEN is not set, please fetch it from 1Password under the key BUILDKITE_GITHUBDOTCOM_TOKEN"
```
Comment (valerybugakov) [inline dev/run-frontend-e2e.sh:10]: I couldn't find `BUILDKITE_GITHUBDOTCOM_TOKEN` in the shared 1Password vault, but `GITHUB_TOKEN` from the `E2E Pipeline Secrets` note works. Should we update either the 1Password note or this message? Or am I looking at the wrong place 🙂

```suggestion
    echo -e "GH_TOKEN is not set, please fetch it from the "E2E Pipeline Secrets" 1Password note under the key GITHUB_TOKEN" 
```

**Review by valerybugakov — APPROVED** [review]
The issue with `GH_TOKEN` is solved!

Still getting the error about `Error: BROWSER must be "chrome" or "firefox"` despite seeing it in the `sg` config. Maybe it's my local issue 🤔

File: dev/run-frontend-e2e.sh (hunk 4 above)
Comment (jhchabran) [reply dev/run-frontend-e2e.sh:10]: I added it, I forgot to do it earlier :)

File: dev/sg/internal/run/command.go [hunk 5]
```diff
@@ -138,6 +189,9 @@ func startCmd(ctx context.Context, dir string, cmd Command, globalEnv map[string
 		stderrWriter = io.MultiWriter(logger, sc.stderrBuf)
 	}
 
+	if cmd.Preamble != "" {
```
Comment (jhchabran) [inline dev/sg/internal/run/command.go:192]: image

**Review by mrnugget — APPROVED** [review]

**Comment by jhchabran** [comment]
@valerybugakov it's running on my side, I have no idea what's the issue with the browser on your machine.

File: dev/sg/internal/run/command.go [hunk 6]
```diff
@@ -110,7 +130,31 @@ func (sc *startedCmd) CapturedStderr() string {
 	return string(sc.stderrBuf.Bytes())
 }
 
-func startCmd(ctx context.Context, dir string, cmd Command, globalEnv map[string]string) (*startedCmd, error) {
+func getSecrets(ctx context.Context, cmd Command) (map[string]string, error) {
+	secretsEnv := map[string]string{}
+
+	client, err := secretmanager.NewClient(ctx)
```
Comment (bobheadxi) [inline dev/sg/internal/run/command.go:136]: @jhchabran why don't we integrate this with secrets store?

Comment (bobheadxi) [reply dev/sg/internal/run/command.go:136]: https://github.com/sourcegraph/sourcegraph/pull/34942
